pkg/agent/manager/comm.go

```diff
@@ -0,0 +1,236 @@
+func (m *manager) newClient(svid *x509.Certificate, key *ecdsa.PrivateKey) (*client, error) {
+	conn, err := m.newGRPCConn(svid, key)
+	if err != nil {
+		return nil, err
+	}
+	nodeClient := node.NewNodeClient(conn)
+
+	return &client{log: m.c.Log, requests: make(chan *node.FetchSVIDRequest), conn: conn, nodeClient: nodeClient}, nil
+}
+
+// newClient adds a new client to the pool and associates it to the specified list of spiffeIDs.
+func (m *manager) newSyncClient(spiffeIDs []string, svid *x509.Certificate, key *ecdsa.PrivateKey) error {
+	// If there is no pool yet, create one.
+	m.mtx.Lock()
+	if m.syncClients == nil {
+		m.syncClients = &clientsPool{clients: map[string]*client{}, m: &sync.Mutex{}}
+	}
+	m.mtx.Unlock()
+
+	client, err := m.newClient(svid, key)
+	if err != nil {
+		return err
+	}
+
+	for _, id := range spiffeIDs {
+		m.syncClients.add(id, client)
+	}
+
+	return nil
+}
+
+func (p *clientsPool) add(spiffeID string, client *client) {
+	// If there is already a connection with the specified spiffeID, close it first.
+	if c := p.get(spiffeID); c != nil {
+		c.close()
+	}
+
+	p.m.Lock()
+	defer p.m.Unlock()
+	p.clients[spiffeID] = client
+}
+
+func (p *clientsPool) get(spiffeID string) *client {
+	p.m.Lock()
+	defer p.m.Unlock()
+	return p.clients[spiffeID]
+}
+
+func (m *manager) getRotationClient() *client {
+	return m.syncClients.get(rotatorTag)
+}
+
+func (m *manager) renewRotatorClient() error {
+	svid, key := m.getBaseSVIDEntry()
+	c, err := m.newClient(svid, key)
+	if err != nil {
+		return err
+	}
+	m.syncClients.add(rotatorTag, c)
+	return nil
+}
+
+// close releases the pool's resources.
+func (p *clientsPool) close() {
+	p.m.Lock()
+	defer p.m.Unlock()
+	for _, c := range p.clients {
+		c.close()
+	}
+}
+
+func (c *client) close() {
+	c.conn.Close()
+}
+
+type update struct {
+	regEntries map[string]*common.RegistrationEntry
+	svids      map[string]*node.Svid
+	lastBundle []byte
+}
+
+func (u *update) String() string {
+	var buffer bytes.Buffer
+	buffer.WriteString("{ regEntries: [")
+	for _, re := range u.regEntries {
+		buffer.WriteString("{ spiffeID: ")
+		buffer.WriteString(re.SpiffeId)
+		buffer.WriteString(", parentID: ")
+		buffer.WriteString(re.ParentId)
+		buffer.WriteString(", selectors: ")
+		buffer.WriteString(fmt.Sprintf("%v", re.Selectors))
+		buffer.WriteString("}")
+	}
+	buffer.WriteString("], svids: [")
+	for key, svid := range u.svids {
+		buffer.WriteString(key)
+		buffer.WriteString(": ")
+		buffer.WriteString(svid.String()[:30])
+		buffer.WriteString(" ")
+	}
+	buffer.WriteString("], lastBundle: ")
+	if u.lastBundle != nil && len(u.lastBundle) > 0 {
+		buffer.WriteString("bytes")
+	} else {
+		buffer.WriteString("none")
+	}
+	buffer.WriteString("}")
+	return buffer.String()
+}
+
+func (c *client) sendAndReceive(r *node.FetchSVIDRequest) (*update, error) {
+	// Enable pipelined access by client to this function.
+	go func() {
+		c.requests <- r
+	}()
+	req := <-c.requests
```

Comment:
I'm trying to understand the purpose of this bit... we write r to c.requests in a goroutine, then immediately read from c.requests... is the goroutine necessary?

Reply:
Sorry about that...I think this is wrong...I will fix this.